extmod/uasyncio/stream.py

```diff
@@ -35,25 +35,29 @@ async def readinto(self, buf):
         return self.s.readinto(buf)
 
     async def readexactly(self, n):
-        r = b""
+        r = []
         while n:
             yield core._io_queue.queue_read(self.s)
             r2 = self.s.read(n)
             if r2 is not None:
                 if not len(r2):
                     raise EOFError
-                r += r2
+                r.append(r2)
                 n -= len(r2)
-        return r
+        if len(r) == 1:
+            return r[0]
+        return b"".join(r)
 
     async def readline(self):
-        l = b""
+        l = []
         while True:
             yield core._io_queue.queue_read(self.s)
             l2 = self.s.readline()  # may do multiple reads but won't block
-            l += l2
-            if not l2 or l[-1] == 10:  # \n (check l in case l2 is str)
-                return l
+            l.append(l2)
+            if not l2 or l2[-1] == 10:  # \n (check l in case l2 is str)
+                if len(l) == 1:
+                    return l[0]
```

Comment:
@dlech I just wonder why we should check that returned value of `s.readline()` is a `str` or not? My second question is how this if statement ( line 57 ) is checking `l2` is str or not by checking `not l2` ?

Reply:
`l2 = self.s.readline()` which can return `str` or `bytes` depending on the type of stream. So if `l` only has one item and it is a `str`, then the `str` is returned. In all other cases, `bytes` is returned. This could be a problem for someone calling `.decode()` on the returned value since it used to always return bytes.

Reply:
@dlech What about `s.read()` ? Does it return str too ?

Reply:
Another question. Assume there's a stream called `s` of type `str` so If you call `s.readline()` it will return `string`. But if you create a `StreamReader` from that stream `StreamReader.readline()` will give you bytes! I think this is not rational that using a stream in uasyncio program causes the stream output type gets changed. Another problem of current (I mean main branch) implementation is If developer wants to get data of `StreamReader.readline()` as string, their should `decode()` it back which means additional `encode()` and `decode()` for nothing.

Reply:
I just want to propose an alternative way to handle this :

- `StreamReader.readline()` returns `str` if underlying stream type is string.
- `StreamReader.readline()` returns `bytes` if underlying stream type is bytes.